lib/classroom_api.rb

```diff
@@ -8,190 +12,229 @@ def initialize(access_token)
-        @log.api_logger.debug "add_facility_id_to_classrooms, error: could not delete records with #{@rooms_in_db} rmrecnbr"
+        @log.api_logger.debug "add_facility_id_to_classrooms, error: API return: #{result['errorcode']} - #{result['error']} for #{facility_id}"
         @debug = true
         return @debug
       end
+      # check if database has rooms that are not in API anymore
+      if @rooms_in_db.present?
+        RoomContact.where(rmrecnbr: @rooms_in_db).delete_all
+        RoomCharacteristic.where(rmrecnbr: @rooms_in_db).delete_all
+        if Room.where(rmrecnbr: @rooms_in_db).delete_all
+          @log.api_logger.info "add_facility_id_to_classrooms, delete #{@rooms_in_db} room(s) from the database"
+        else
+          @log.api_logger.debug "add_facility_id_to_classrooms, error: could not delete records with #{@rooms_in_db} rmrecnbr"
+          @debug = true
+          return @debug
+        end
+      end
+    rescue StandardError => e
+      # example: Errno::ETIMEDOUT: Operation timed out - user specified timeout
+      @log.api_logger.debug "add_facility_id_to_classrooms, error: API return: #{e.message}"
+      @debug = true
     end
     return @debug
   end
 
 
   def get_classrooms_list
-    result = {'success' => false, 'errorcode' => '', 'error' => '', 'data' => {}}
-    url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/Classrooms?BuildingID=1005046")
-
-    http = Net::HTTP.new(url.host, url.port)
-    http.use_ssl = true
-    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
-
-    request = Net::HTTP::Get.new(url)
-    request["x-ibm-client-id"] = "#{Rails.application.credentials.um_api[:buildings_client_id]}"
-    request["authorization"] = "Bearer #{@access_token}"
-    request["accept"] = 'application/json'
-
-    response = http.request(request)
-    response_json = JSON.parse(response.read_body)
-    if response_json['errorCode'].present?
-      result['errorcode'] = response_json['errorCode']
-      result['error'] = response_json['errorMessage']
-    else
-      result['success'] = true
-      result['data'] = response_json['Classrooms']['Classroom']
+    begin
+      result = {'success' => false, 'errorcode' => '', 'error' => '', 'data' => {}}
+      classrooms = []
+      start_index = 0
+      count = 1000
+      next_page = true
+      while next_page do
+        url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=#{start_index}&$count=#{count}")
+
+        http = Net::HTTP.new(url.host, url.port)
+        http.use_ssl = true
+        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+
+        request = Net::HTTP::Get.new(url)
+        request["x-ibm-client-id"] = "#{Rails.application.credentials.um_api[:buildings_client_id]}"
+        request["authorization"] = "Bearer #{@access_token}"
+        request["accept"] = 'application/json'
+
+        response = http.request(request)
+        link = response.to_hash["link"].to_s
+        if link.include? "rel=next"
```

Comment:
@britaumich - We can extract next link using regex as follows:

`link[/<(\S+)>; rel=next/,1]`

I hope this helps.

Reply:
Yes, we can, but I don't understand why we want to do that. the URL already exists. It was created on line 102:
`url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=#{start_index}&$count=#{count}")`
The only thing that should be changed in the url variable is start_index.
Why 
`url = URI(link[/<(\S+)>; rel=next/,1])`
is better than
`start_index += count` ?

Their documentation explains how their pagination works, not how it should be used in Rybu.

Reply:
The document does mention about how to use it.

"_Users can directly use the urls in the link header to navigate to the next set of results._"

Reference - https://dir.api.it.umich.edu/guides/migration-to-denodo

Reply:
So we have two approaches here and I am suggesting this new solution based on the preferred approach mentioned in the document. I will leave the final decision to you.

Reply:
They said 'uses can' not 'users should'. They could choose different wards to describe that. For example 'The returned header includes the url...' You don't need to follow every word someone wrote. 
You did not answer my question:
Why
`url = URI(link[/<(\S+)>; rel=next/,1])`
is better than
`start_index += count` ?

It's clear from my code what happens if start_index is increased by count. This is how pagination works: you request the next 'count' number of records.
It's not clear what this code `url = URI(link[/<(\S+)>; rel=next/,1])` means.

Reply:
1) `url = URI(link[/<(\S+)>; rel=next/,1])` approach is better because you get direct link from the response. With the other approach, you are building the link.

2) This is normal regex  - `/<(\S+)>; rel=next/ ` and there is nothing complex about it.

Reply:
I had to build the url to call the API the first time. 
I understand regex, but it's not clear - just looking at the code - why you need this expression. And using 
start_index += count is simple and logical.

Reply:
Why do you want to build something when it is already available in the API response ?

Reply:
Exactly - in response. But to get the response you need to create the url variable. So, why do you need to recreate it?
```
def get_classrooms_list
    begin
      result = {'success' => false, 'errorcode' => '', 'error' => '', 'data' => {}}
      classrooms = []
      start_index = 0
      count = 1000
      next_page = true
      while next_page do
        url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=#{start_index}&$count=#{count}")

        http = Net::HTTP.new(url.host, url.port)
        http.use_ssl = true
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE

        request = Net::HTTP::Get.new(url)
        request["x-ibm-client-id"] = "#{Rails.application.credentials.um_api[:buildings_client_id]}"
        request["authorization"] = "Bearer #{@access_token}"
        request["accept"] = 'application/json'

        response = http.request(request)
        link = response.to_hash["link"].to_s
        if link.include? "rel=next"
          start_index += count
        else
          next_page = false
        end
        response_json = JSON.parse(response.read_body)
        if response.code == OK_CODE
          result['success'] = true
          classrooms += response_json['Classrooms']['Classroom']
        elsif response_json['errorCode'].present?
          result['errorcode'] = response_json['errorCode']
          result['error'] = response_json['errorMessage']
          next_page = false
        else 
          result['errorcode'] = "Unknown error"
          next_page = false
        end
      end
      result['data'] = classrooms
    rescue StandardError => e
      result['errorcode'] = "Exception"
      result['error'] = e.message
    end
    return result
  end
```

Reply:
I did not understand your comment about creating variable.

Following is the solution with regex approach. I deleted variables **_count_** and **_start_index_** since we are not building url with this approach. I have marked the modified code in bold and italics.

`def get_classrooms_list
    begin
      result = {'success' => false, 'errorcode' => '', 'error' => '', 'data' => {}}
      classrooms = []
      **_url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=0&$count=1000")_**
      next_page = true
      while next_page do
        http = Net::HTTP.new(url.host, url.port)
        http.use_ssl = true
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE

        request = Net::HTTP::Get.new(url)
        request["x-ibm-client-id"] = "#{Rails.application.credentials.um_api[:buildings_client_id]}"
        request["authorization"] = "Bearer #{@access_token}"
        request["accept"] = 'application/json'

        response = http.request(request)
        link = response.to_hash["link"].to_s

        **_url = link[/<(\S+)>; rel=next/,1]
        if !url
          next_page = false
        end_**

        response_json = JSON.parse(response.read_body)
        if response.code == OK_CODE
          result['success'] = true
          classrooms += response_json['Classrooms']['Classroom']
        elsif response_json['errorCode'].present?
          result['errorcode'] = response_json['errorCode']
          result['error'] = response_json['errorMessage']
          next_page = false
        else 
          result['errorcode'] = "Unknown error"
          next_page = false
        end
      end
      result['data'] = classrooms
    rescue StandardError => e
      result['errorcode'] = "Exception"
      result['error'] = e.message
    end
    return result
end`

Reply:
`def get_classrooms_list
    begin
      result = {'success' => false, 'errorcode' => '', 'error' => '', 'data' => {}}
      classrooms = []
      **_url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=0&$count=1000")_**
      next_page = true
      while next_page do
        http = Net::HTTP.new(url.host, url.port)
        http.use_ssl = true
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE

        request = Net::HTTP::Get.new(url)
        request["x-ibm-client-id"] = "#{Rails.application.credentials.um_api[:buildings_client_id]}"
        request["authorization"] = "Bearer #{@access_token}"
        request["accept"] = 'application/json'

        response = http.request(request)
        link = response.to_hash["link"].to_s

        **_url = link[/<(\S+)>; rel=next/,1]
        if !url
          next_page = false
        end_**

        response_json = JSON.parse(response.read_body)
        if response.code == OK_CODE
          result['success'] = true
          classrooms += response_json['Classrooms']['Classroom']
        elsif response_json['errorCode'].present?
          result['errorcode'] = response_json['errorCode']
          result['error'] = response_json['errorMessage']
          next_page = false
        else 
          result['errorcode'] = "Unknown error"
          next_page = false
        end
      end
      result['data'] = classrooms
    rescue StandardError => e
      result['errorcode'] = "Exception"
      result['error'] = e.message
    end
    return result
end`

Reply:
First, the correct code will be 
```url = URI(link[/<(\S+)>; rel=next/,1])```, not `url = link[/<(\S+)>; rel=next/,1]`

Second, I still prefer my code.

Reply:
`url = link[/<(\S+)>; rel=next/,1]` is correct. 

Please check my code again.

Reply:
I would also like to hear Rick's opinion regarding both approaches.